docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,921 @@
+
+### Divide and conquer
+
+Considering the complexity of this topic, a phased approach is required to design and implement the solution for in-place upgrades.
+
+The main goal of the first iteration of this proposal is to make it possible for Cluster API users to start experimenting usage of in-place upgrades, so we can gather feedback and evolve to the next stage.
+
+This iteration will focus on implementing the machinery required to interact with update extensions, while users facing changes in the API types are deferred to follow up iterations.
+
+### Tenets
+
+#### Same UX
+
+Cluster API user experience MUST be the same when using default, immutable updates or when using external update extensions: e.g. in order to trigger a MachineDeployment rollout, you have to rotate a template, etc.
+
+#### Fallback to Immutable rollouts
+
+If external update extensions can not cover the totality of the desired changes, CAPI WILL defer to Cluster API’s default, immutable rollouts. This is important for a couple of reasons:
+
+* It allows to implement custom rollout strategies incrementally, without the need to cover all use cases up-front.
+* There are cases when replacing the machine will always be necessary:
+    * When it is not possible to recover the machine, e.g. hardware failure.
+    * When the user determines that recovering the machine is too complex/costly vs replacing it. 
+    * Automatic machine remediation (unless you use external remediation strategies)
+
+#### Clean separation of concern
+
+It is the responsibility of the extension to decide if it can perform changes in-place and to perform these changes on a single machine. If the extension decides that it cannot perform changes in-place, CAPI will fall back to rollout.
+
+The responsibility to determine which machine should be rolled out as well as the responsibility to handle rollout options like MaxSurge/MaxUnavailable will remain on the controllers owning the machine (e.g. KCP, MD controller). 
+
+### Goals
+
+- Enable the implementation of pluggable update extensions.
+- Allow users to update Kubernetes clusters using pluggable External Update Extension.
+- Support External Update Extensions for both Control Plane (KCP or others) and MachineDeployment controlled machines.
+
+### Non-Goals/Future work
+
+- To provide rollbacks in case of an in-place update failure. Failed updates need to be fixed manually by the user on the machine or by replacing the machine.
+- Introduce any changes to KCP (or any other control plane provider), MachineDeployment, MachineSet, Machine APIs.
+- Maintain a coherent user experience for both rolling and in-place updates.
+- Allow in-place updates for single-node clusters without the requirement to reprovision hosts (future goal).
```

Comment:
I think this comment is still valid

> What about the OnDelete strategy of MDs?
> Probably we just shouldn't try in-place if OnDelete is configured? (so maybe it's a non-goal?)
> https://github.com/kubernetes-sigs/cluster-api/pull/11029#discussion_r2007295843

(just opened a new comment to make it easier to deal with, considering GitHub UI limitations...)

Reply:
Clarifying, are you suggesting we add "Support MDs with OnDelete set" as a non-goal?

Reply:
In general it would be good to know what our plan is for MDs that use OnDelete. On a high-level it sounds fundamentally incompatible with the idea of in-place updates

Reply:
But I guess we'll see during implementation, fine for now